guppylang/std/qsystem/__init__.py

```diff
@@ -74,6 +76,52 @@ def reset(q: qubit) -> None: ...
 def qfree(q: qubit @ owned) -> None: ...
 
 
+@guppy.hugr_op(quantum_op("LazyMeasureLeaked", ext=QSYSTEM_EXTENSION))
+@no_type_check
+def _measure_leaked(q: qubit @ owned) -> Future[int]:
+    """Measure the quibit or return 2 if it is leaked."""
+
+
+@guppy
+@no_type_check
+def measure_leaked(q: qubit @ owned) -> "MaybeLeaked":
+    fm = _measure_leaked(q)
+    return MaybeLeaked(fm)
+
+
+@guppy.struct
+@no_type_check
+class MaybeLeaked:
+    """A class representing a measurement that may have leaked.
+
+    This is used to represent the result of `measure_leaked`, which can either
+    return a boolean measurement result or indicate that the qubit has leaked.
+    """
+
+    _measurement: Future[int]  # type: ignore[type-arg]
+
+    @guppy
+    @no_type_check
+    def is_leaked(self: "MaybeLeaked") -> bool:
+        """Check if the measurement indicates a leak."""
+        return self._measurement.copy().read() == 2
+
+    @guppy
+    @no_type_check
+    def to_result(self: "MaybeLeaked @ owned") -> Option[bool]:
+        """Get the measurement result if not leaked."""
```

Comment:
Maybe something like this would be more accurate?
```suggestion
        """Returns the measurement result or `nothing` if leaked."""
```